storage/spanner.go

```diff
@@ -0,0 +1,984 @@
+}
+
+func (s Storage) performPutOperation(ctx context.Context, t *spanner.ReadWriteTransaction, table string, m map[string]interface{}) error {
+	ddl := models.TableDDL[table]
+	for k, v := range m {
+		t, ok := ddl[k]
+		if t == "BYTES(MAX)" && ok {
+			ba, err := json.Marshal(v)
+			if err != nil {
+				return errors.New("ValidationException", err)
+			}
+			m[k] = ba
+		}
+	}
+
+	mutation := spanner.InsertOrUpdateMap(table, m)
+	mutations := []*spanner.Mutation{mutation}
+	err := t.BufferWrite(mutations)
+	if e := errors.AssignError(err); e != nil {
+		return e
+	}
+	return nil
+}
+
+// SpannerBatchPut - this insert or update data in batch
+func (s Storage) SpannerBatchPut(ctx context.Context, table string, m []map[string]interface{}) error {
+	mutations := make([]*spanner.Mutation, len(m))
+	ddl := models.TableDDL[changeTableNameForSP(table)]
+	table = changeTableNameForSP(table)
+	for i := 0; i < len(m); i++ {
+		for k, v := range m[i] {
+			t, ok := ddl[k]
+			if t == "BYTES(MAX)" && ok {
+				ba, err := json.Marshal(v)
+				if err != nil {
+					return errors.New("ValidationException", err)
+				}
+				m[i][k] = ba
+			}
+		}
+		mutations[i] = spanner.InsertOrUpdateMap(table, m[i])
+	}
+	_, err := s.getSpannerClient(table).Apply(ctx, mutations)
+	if err != nil {
+		return errors.New("ResourceNotFoundException", err.Error())
+	}
+	return nil
+}
+
+// SpannerDelete - this will delete the data
+func (s Storage) SpannerDelete(ctx context.Context, table string, m map[string]interface{}, eval *models.Eval, expr *models.UpdateExpressionCondition) error {
+	_, err := s.getSpannerClient(table).ReadWriteTransaction(ctx, func(ctx context.Context, t *spanner.ReadWriteTransaction) error {
+		tmpMap := map[string]interface{}{}
+		for k, v := range m {
+			tmpMap[k] = v
+		}
+		if len(eval.Attributes) > 0 || expr != nil {
+			status, err := evaluateConditionalExpression(ctx, t, table, tmpMap, eval, expr)
+			if err != nil {
+				return err
+			}
+			if !status {
+				return errors.New("ConditionalCheckFailedException", tmpMap, expr)
+			}
+		}
+		tableConf, err := config.GetTableConf(table)
+		if err != nil {
+			return err
+		}
+		table = changeTableNameForSP(table)
+
+		pKey := tableConf.PartitionKey
+		pValue, ok := tmpMap[pKey]
+		if !ok {
+			return errors.New("ResourceNotFoundException", pKey)
+		}
+		var key spanner.Key
+		sKey := tableConf.SortKey
+		if sKey != "" {
+			sValue, ok := tmpMap[sKey]
+			if !ok {
+				return errors.New("ResourceNotFoundException", pKey)
+			}
+			key = spanner.Key{pValue, sValue}
+
+		} else {
+			key = spanner.Key{pValue}
+		}
+
+		mutation := spanner.Delete(table, key)
+		err = t.BufferWrite([]*spanner.Mutation{mutation})
+		if e := errors.AssignError(err); e != nil {
+			return e
+		}
+		return nil
+	})
+	return err
+}
+
+// SpannerBatchDelete - this delete the data in batch
+func (s Storage) SpannerBatchDelete(ctx context.Context, table string, keys []map[string]interface{}) error {
```

Comment:
Same as above: `SpannerBatchDelete` and `SpannerDelete` should be combined to share the same logic and strategy.

Reply:
@rosspatil